log4j-api/src/main/java/org/apache/logging/log4j/util/PropertiesUtil.java

```diff
@@ -517,45 +532,64 @@ private Environment(final PropertySource propertySource) {
                             ServiceLoader.load(PropertySource.class, PropertiesUtil.class.getClassLoader()),
                             LOGGER)
                     .forEach(sources::add);
+            sources.sort(Comparator.INSTANCE);
 
             reload();
         }
 
-        /**
-         * Allow a PropertySource to be added.
-         * @param propertySource The PropertySource to add.
-         */
-        public void addPropertySource(final PropertySource propertySource) {
-            sources.add(propertySource);
+        private void addPropertySource(final PropertySource propertySource) {
+            try {
+                sourceWriteLock.lock();
+                if (!sources.contains(propertySource)) {
+                    sources.add(propertySource);
+                    sources.sort(Comparator.INSTANCE);
```

Comment:
I thought we agreed to only sort while iterating, but, AFAICT, we are doing something else. Any particular reason for this choice?

Reply:
`ConcurrentSkipListSet` can not be used, since it needs a `Comparator` (either provided or the class must be `Comparable`).

This is why we use a `List` to store the property sources. Sorting the list allows the use of `List.forEach`, which is less expensive than streams.

Reply:
I am not able to follow. What is wrong with *only* doing `sources.stream().sorted(Comparator.INSTANCE).forEach()`?

Reply:
I didn't run any benchmark, but I would say that it is slower than `sources.forEach()`.

Reply:
How often `reload()` is called? Is the performance difference in the big picture notable enough to warrant the extra optimization? If not, I am inclined to go with the simplest approach.

I am not a big fan on these kind of discussions. Please address this conversation as you see fit, resolve it, and let's merge this PR.

Reply:
I ran some benchmarks of these methods:
```java

public final List<PropertySource> arrayList = new ArrayList<>();
public final Set<PropertySource> skipSet = ConcurrentHashMap.newKeySet();

@Benchmark
public void arrayList(final Blackhole bh) {
    sourceReadLock.lock();
    try {
        for (int i = 0; i < count; i++) {
            bh.consume(arrayList.get(i).getProperty("KEY"));
        }
    } finally {
        sourceReadLock.unlock();
    }
}

@Benchmark
public void skipList(final Blackhole bh) {
    skipSet.stream().sorted(comparator).forEach(source -> bh.consume(source.getProperty("KEY")));
}
```

with 8 property sources and the following results:

```
Benchmark                  (count)  Mode  Cnt    Score   Error  Units
StreamBenchmark.arrayList        8  avgt   25   20,696 ± 0,188  ns/op
StreamBenchmark.skipList         8  avgt   25  111,643 ± 4,113  ns/op
```

Assuming a 100 `getProperty` calls during initialization we are talking about a difference between `~2 µs` and `~10 µs`. It's probably not worth the additional code size of the `ArrayList`.